src/OrleansRuntime/Core/Dispatcher.cs

```diff
@@ -597,32 +598,60 @@ private void ResendMessageImpl(Message message, ActivationAddress forwardingAddr
         #region Send path
 
         /// <summary>
-        /// Send an outgoing message
+        /// Send an outgoing message, may complete synchronously
         /// - may buffer for transaction completion / commit if it ends a transaction
         /// - choose target placement address, maintaining send order
         /// - add ordering info and maintain send order
         /// 
         /// </summary>
         /// <param name="message"></param>
         /// <param name="sendingActivation"></param>
-        public async Task AsyncSendMessage(Message message, ActivationData sendingActivation = null)
+        public Task AsyncSendMessage(Message message, ActivationData sendingActivation = null)
         {
-            try
-            {
-                await AddressMessage(message);
-                TransportMessage(message);
-            }
-            catch (Exception ex)
+            Action<Exception> onAddressingFailure = ex =>
             {
                 if (ShouldLogError(ex))
                 {
-                    logger.Error(ErrorCode.Dispatcher_SelectTarget_Failed,
-                        String.Format("SelectTarget failed with {0}", ex.Message),
-                        ex);
+                    logger.Error(ErrorCode.Dispatcher_SelectTarget_Failed, $"SelectTarget failed with {ex.Message}", ex);
                 }
+
                 MessagingProcessingStatisticsGroup.OnDispatcherMessageProcessedError(message, "SelectTarget failed");
                 RejectMessage(message, Message.RejectionTypes.Unrecoverable, ex);
+            };
```

Comment:
What's the purpose of allocating this delegate with closures for every call? Can't you just define an actual method that you call from both places where failures occur, but doesn't allocate unnecessarily?